README.md
# Hostathome : l'auto-hébergement facile

Host@home est un simple script shell permettant d'installer le plus
simplement possible son serveur à la maison et donc s'*auto-héberger*.
Il fonctionne avec [debian](http://debian.org) wheezy,
[raspbian](http://www.raspbian.org) et sans doute sur ses dérivées.

L'objectif est le suivant : n'importe qui, avec peu de connaissances en
réseau/serveur/linux devrait pouvoir auto-héberger chez soi les services
dont il a besoin, afin de favoriser un internet décentralisé et les
libertés de chacun. Hostathome doit alors permettre : 

- D'être facile d'utilisation,
- Proposer des programmes légers pouvant tourner sur un maximum de
  machines même de récupération,
- Fournir une configuration de base pour les services,
- Sécuriser au maximum le serveur, ce qui ne doit pas être une prise de
  tête pour le débutant,
- Être conçu de façon à pouvoir ajouter facilement le support de
  nouveaux services.

Pour une utilisation optimale de votre serveur, il est pratique d'avoir
un nom de domaine. Pour plus d'infos, voir [cette partie](#DNS).

- [Page
  "officielle"](http://yeuxdelibad.net/Programmation/Hostathome.html)
- [Hostathome sur github](https://github.com/Ikse/hostathome)

## Sommaire

- [Services proposés](#services)
- [Installation](#installation)
- [Utilisation](#utilisation)
- [Configuration](#configuration)
- [Commentaires](#commentaires)
- [Contribuez](#contribuez)
- [Application non supportée](#nonsupportee)
- [Astuces](#DNS)
- [Liens](#liens)
- [Captures d'écran](#scrot)

## Services proposés <a id="services"></a>
Actuellement, le script permet d'installer et d'*auto-héberger* les
services listés ci-dessous. D'autres viendront à l'avenir selon vos
suggestions. Notez que toutes les "webapps", services accessibles via
navigateur internet sont installés avec nginx comme serveur http.

- un site web via http et/ou https (nginx),
- Un serveur de partage de fichiers avec sftp,
- Serveur VPN avec [OpenVPN](http://openvpn.net),
- Installation de [owncloud](http://owncloud.org), 
- Installation de [Pydio](http://pyd.io) anciennement Ajaxplorer, pour
  l'hébergement de fichiers,
- Un serveur de courriel avec [postfix](http://www.postfix.org) + postgrey + [dovecot](http://www.dovecot.org),
- Un webmail, pour consulter son courrier via navigateur
  ([squirrelmail](http://squirrelmail.org), [rainloop](http://rainloop.net)) ou [roundcube](http://roundcube.net/),
- Un serveur xmpp (jabber) avec [prosody](https://prosody.im),
- Installation de [dropcenter](http://projet.idleman.fr/dropcenter/)
  pour partager les fichiers,
- Installation de [pluxml](http://www.pluxml.org/) un moteur de
  blog/CMS,
- Installation de
  [blogoText](http://lehollandaisvolant.net/blogotext/fr/) un moteur de
  blog léger,
- Installation de [Kriss](http://tontof.net/kriss/feed/) un lecteur de
  flux rss très simple, ou de 
  [Tiny Tiny RSS](http://tt-rss.org/redmine/projects/tt-rss/wiki),
- Installation d'un relais [tor](https://www.torproject.org) pour
  participer au réseau,
- Installation de [wallabag](https://www.wallabag.org/) pour sauver des
  pages à lire plus tard,
- Installation de [pico](http://pico.dev7studios.com/), un CMS très
  simple + son éditeur en ligne,
- Installation de [wallabag](https://www.wallabag.org/) pour sauver des
  pages à lire plus tard,
- Installation de [jyraphe](http://home.gna.org/jyraphe/), pour déposer
  des fichiers trop lourds à envoyer par mail, et donner le lien de
  téléchargement à un ami,
- Installation de forums ([NoNonsenseForum](http://camendesign.com/nononsense_forum) un forum minimaliste, [FluxBB](http://fluxbb.org/), [phpBB](https://www.phpbb.com/)),
- Installation de
  [Shaarli](http://sebsauvage.net/wiki/doku.php?id=php:shaarli) le
  delicious-like de sebsauvage,
- Installation de
  [ZeroBin](http://sebsauvage.net/wiki/doku.php?id=php:zerobin), le pastebin de sebsauvage,
- Installation d'une seedbox torrent
  ([rtorrent](http://libtorrent.rakshasa.no/) et
  [rutorrent](https://code.google.com/p/rutorrent/)),
- Installation de wikis comme [DokuWiki](https://www.dokuwiki.org/dokuwiki) ou
  [MediaWiki](https://www.mediawiki.org),
- Installation de gestionnaires de listes de tâches : 
[mytinytodo](http://www.mytinytodo.net), 
  [jotter](http://www.yosko.net/article31/jotter-notebook-manager), 
  [kanboard](http://kanboard.net),
- Installation d'un serveur DNS [unbound](http://www.unbound.net),
- Installation de [monitorix](http://www.monitorix.org) pour superviser votre serveur.

## Installation <a id="installation"></a>

Récupérez la dernière archive
[hostathome-last.tar.gz](http://yeuxdelibad.net/DL/hostathome-last.tar.gz) et décompressez-la avec la commande  : 

    wget http://yeuxdelibad.net/DL/hostathome-last.tar.gz
    tar -xvf hostathome-last.tar.gz
    cd hostathome
    ./hostathome-dialog.sh

Pour profiter de la toute dernière version du script,
vus devrez cependant avoir le paquet *git* d'installé
pour pouvoir accéder à la commande `git`. Récupérez-le avec les commandes
suivantes : 

    git clone https://github.com/Ikse/hostathome.git
  

##Utilisation <a id="utilisation"></a>
Lancez le script simplement `./hostathome-dialog.sh` avec les droits
superutilisateur, puis suivez les
instructions. Le script vous posera alors quelques questions pour
connaître l'emplacement géographique du serveur (certificats ssl
obligent), le dossier où vous souhaitez installer vos services webs, le
[nom de domaine](#DNS) pour accéder à votre site...

Vous pouvez aussi utiliser le script de façon automatisée afin de
simplifier des installations multiples, sur plusieurs serveurs par
exemple.

Pour cela, lancez le script ainsi `./hostathome-dialog.sh config.cfg` où
config.cfg est le fichier de configuration contenant les instructions.
Un exemple est fournit avec le script, il suffit de regarder dans
`hah.cfg`.

Dans tous les cas, **pensez à lire le fichier RAPPORT** créé par le
script après l'avoir utilisé.


##Configuration <a id="configuration"></a>
Certains services nécessitent une configuration manuelle de votre part.
Ces informations sont indiquées dans le fichier RAPPORT généré lors de
l'utilisation de hostathome.

Pour ouvrir les ports de votre routeur, cela dépend du modèle fourni par votre FAI. Plus d'informations [par
ici](https://craym.eu/tutoriels/utilitaires/ouvrir_les_ports_de_sa_box.html)

Voici donc les opérations à effectuer après l'installation de certains
services : 

- Site internet (nginx) : ouvrir et rediriger les ports 80 et 443
- Tor : ouvrir et rediriger le port 9001
- Courriel (postfix) : 
        - ouvrir et rediriger les ports 25, 587, 143, 993.
        - Ajoutez les champs DNS indiqués par le script dans les Zones DNS de votre registrar. La procédure dépend de
        ce dernier. Chez OVH par exemple, cela se passe dans le manager.
        - Pour ajouter une nouvelle adresse de courriel, créez un nouvel utilisateur. 
        Par exemple, pour avoir `toto@$votredomaine.com`, créez l'utilisateur `toto` avec la commande
                `adduser toto`
        - Vous pouvez ensuite configurer un client email. Pour la
          réception, il faut choisir IMAP avec SSL port 993. Pour
          l'envoi, vous pouvez choisir soit le port 25 habituel, ou bien
          le port 587 si votre FAI bloque le 25.
- Tous les services sur https (blogotext, pluxml...) : rediriger et
  ouvrir le port 443
- XMPP : ajouter les zones DNS indiquées chez votre registrar, et ouvrir/rediriger les ports 5222 et 5269. Il vous suffira ensuite d'utiliser un  [client](https://fr.wikipedia.org/wiki/Liste_de_clients_XMPP) comme [gajim](http://gajim.org)
- Le webmail rainloop : Allez sur votre site dans un navigateur à
  l'adresse https://votredomaine?admin (il faut juste rajouter `?admin`,
  puis connectez-vous à l'interface de configuration avec les
  identifiants suivants : 
    - Identifiant : admin
    - Mot de passe (à changer au plus vite) : 12345
- NoNonsenseForum peut être configuré en éditant le fichier config.php
  présent dans le dossier d'installation
- kanboard : les identifiants par défaut sont admin/admin
- Pour la seedbox avec rtorrent, vos téléchargements seront acessibles
  en ajoutant "/downloads" à l'adresse de l'interface web. Exemple
  "https://monrtorrent.domaine.com/downloads",
- Pour augmenter la taille limite des fichiers en upload pour php (Dropcenter, blogotext, owncloud, jyraphe, mediawiki, dokuwiki, pydio), modifiez la ligne suivante dans le fichier de configuration nginx, par exemple/etc/nginx/conf.d/dropcenter.conf

    `client_max_body_size 1000M;`

ainsi que dans le fichier /etc/php5/fpm/php.ini

    `upload_max_filesize = 1000M`
    `post_max_size = 1000M;`


##Vos commentaires <a id="commentaires"></a>
N'hésitez pas à rapporter les bugs, donner des
idées, proposer de nouveaux services ou à
demander de l'aide soit [en me contactant
directement](/Divers/Contact.html) ou bien en ouvrant un fil [sur le
forum](http://forum.yeuxdelibad.net).

##Contribuez! <a id="contribuez"></a>
Il est normalement facile d'ajouter des services à ce script. 
N'hésitez donc pas à contribuer en [me contactant](/Divers/Contact.html)
pour obtenir les droits d'écriture sur le dépôt.

Vous pouvez aussi participer [au dépôt
github](https://github.com/Ikse/hostathome)

##Mon application préférée n'est pas supportée <a id="nonsupportee"></a>

Vous pouvez alors contribuer.

Sinon, peut-être fait-elle partie de cette liste : 

- [seafile](http://www.seafile.com) est excellent pour partager les documents.
  Cependant, il n'est disponible que pour certaines architectures, ce
  qui ne convient pas à la "philosophie" du script.
- [Cozycloud](http://cozy.io) -> L'installation et la mise à jour est un vrai bazar à l'heure actuelle (23/04/14),

## Astuce <a id="DNS"></a>

Pour contacter votre serveur, il est bien plus pratique de taper
"machinbidule.net" que "145.332.210.45" qui est l'ip du serveur ici.
(j'invente hein...).  "machinbidule.net" s'appelle un nom de domaine, ou
enregistrement DNS : en gros, c'est un truc qui dit "tu veux aller à
machinbidule.net? En fait, c'est à l'adresse ip 145.332.210.45". Bien
sûr, ça se passe entre les machines du réseau, et c'est bien plus
pratique pour nous les humains.

Pour en bénéficier, vous pouvez en louer chez des registres comme
[Gandi](www.gandi.net) ou [OVH](www.ovh.com).

Cependant pour commencer je vous conseille des gratuits comme
[no-ip](www.no-ip.com).

Dans tous les cas, la lecture de [cette
page](http://wiki.auto-hebergement.fr/domaines/obtenir_un_domaine) n'est
pas à exclure, c'est sur le wiki d'auto-hébergement.

Petite astuce : il est très pratique d'avoir des sous-domaines. Par
exemple, votre domaine est *mondomaine.com*. Afin de mieux gérer vos
services, vous pouvez créer des sous-domaines du type
*owncloud.mondomaine.com*, *forum.mondomaine.com* ...

Pour cela, créez un champ DNS de type CNAME. Faites-le pointer vers le
nom de domaine principal *mondomaine.com*, et donnez-lui le nom que vous
voulez, par exemple *forum.mondomaine.com*.


## Liens divers <a id="liens"></a>
- [Installation détaillée d'un
  serveur](/Logiciel-libre/Installation_et_securisation_d_un_serveur_auto-heberge.html)
- [YUNoHost](http://yunohost.org) facilite aussi l'auto-hébergement
- [Hostathome sur
  linuxfr](https://linuxfr.org/users/thuban-0/journaux/host-home-faciliter-l-auto-hebergement)

Passez faire un tour sur le wiki
[d'auto-hébergement](http://wiki.auto-hebergement.fr), notamment pour
acquérir un [nom de
domaine](http://wiki.auto-hebergement.fr/domaines/obtenir_un_domaine),
et configurer [votre
routeur et parefeu](http://wiki.auto-hebergement.fr/r%C3%A9seau/routeur/configuration).

##Captures d'écran <a id="scrot"></a>

Petites captures d'écran pour vous montrer : 

[![hostathome1-thumb.jpg](/Images/hostathome1-thumb.jpg)](/Images/hostathome1.png)
[![hostathome2-thumb.jpg](/Images/hostathome2-thumb.jpg)](/Images/hostathome2.png)
[![hostathome3-thumb.jpg](/Images/hostathome3-thumb.jpg)](/Images/hostathome3.png)


DEBIAN/changelog
hostathome (0.8) unstable; urgency=low

    * Initial release

-- Xavier Cartron <thuban@yeuxdelibad.net>, 16 Oct. 2014


hostathome-dialog.sh
#!/bin/bash
CURDIR="$(echo $(dirname $(readlink -f $0)))"
#debian
if [ -f /usr/share/hostathome/hah-engine.sh ]; then
    . /usr/share/hostathome/hah-engine.sh
else
    . "$CURDIR/hah-engine.sh"
fi

# gettext
#. gettext.sh
#export TEXTDOMAIN=hostathome
#export TEXTDOMAINDIR=/usr/share/locale

# Global variables
SSLCOUNTRY=""
SSLSTATE=""
SSLLOCATION=""
SSLORGANISATION=""
SSLCOMMONNAME=""
SSLemailAddress=""
SSLFIRSTRUN=""

dgetinfo() {
    #dgetinfo <variable> <message> <titre> <default>
    local rep=""
    while [ -z "$rep" ]; do
        dialog --title "$3" --inputbox "$2" 20 60 "$4" 2> "$tmpfile"
        rep="$(<"$tmpfile")"
    done
    assign "$1" <<< $rep 
}

info_ssl() {
    while [ -z "$SSLCOUNTRY" ]; do
        dialog --title "Configuration de SSL" --inputbox "Code du pays (ex : FR)" 20 60 2> "$tmpfile"
        SSLCOUNTRY="$(<"$tmpfile")"
    done
    while [ -z "$SSLSTATE" ]; do
        dialog --title "Configuration de SSL" --inputbox "Nom du pays? (ex : France)" 20 60 2> "$tmpfile"
        SSLSTATE="$(<"$tmpfile")"
    done
    while [ -z "$SSLLOCATION" ]; do
        dialog --title "Configuration de SSL" --inputbox "Ville? (ex : Paris)" 20 60 2> "$tmpfile"
        SSLLOCATION="$(<"$tmpfile")"
    done
    while [ -z "$SSLemailAddress" ]; do
        dialog --title "Configuration de SSL" --inputbox "Votre addresse de courriel?" 20 60 2> "$tmpfile"
        SSLemailAddress="$(<"$tmpfile")"
    done
    if [ ! -z "$SSLFIRSTRUN" ]; then
        dialog --title "Configuration de SSL" --inputbox "Nom de l'organisation? (peut être laissé vide)" 20 60 2> "$tmpfile"
        SSLORGANISATION="$(<"$tmpfile")"
        dialog --title "Configuration de SSL" --inputbox "Votre nom? (peut être laissé vide)" 20 60 2> "$tmpfile"
        SSLCOMMONNAME="$(<"$tmpfile")"
    fi
    SSLFIRSTRUN="non"
}

info_nginx() {
# met la commande de configuration de nginx dans $1
    # dohttp_nginx <nom d'hote> </emplacement/du/site> <php 0 pour oui/1 pour non> <SSL (1:N, 2:O, 3:O avec redirection)>
    local NOMDHOTE=""
    local ROOTOFHTTP=""
    local DOPHP=""
    local SSL=""

    # Nom de domaine
    dgetinfo NOMDHOTE "Quel est votre nom de domaine (sans http://)? (ex : mondomaine.com) " "Configuration de nginx"

    # Racine du site
    dgetinfo ROOTOFHTTP "Dossier contenant votre site? (ex : /srv/www/supersite) " "Configuration de nginx"

    # PHP
    dialog --title "Configuration de nginx" --yesno "Voulez-vous utiliser php? " 20 60
    DOPHP=$?

    # SSL
    while ! [[ $SSL =~ ^[0-9]+$ ]]; do
        dialog --backtitle "SSL" \
        --radiolist 'Voulez-vous installer le support de ssl? (https)' 20 60 5 \
        "1" "Non, http seul" on\
        "2" "Oui, https seul" off\
        "3" "oui, avec http redirigé vers https" off 2> "$tmpfile"
        SSL="$(<"$tmpfile")"
    done
    if [ "$SSL" = "2" -o "$SSL" = "3" ]; then
        info_ssl
    fi
    assign "$1" <<< "dohttp_nginx $NOMDHOTE $ROOTOFHTTP $DOPHP $SSL"
}

info_postfix() {
    local NOMDHOTE="$1"
    local DOMAIN="$2"

    dialog --msgbox "Un serveur de courriel a besoin de 2 champs DNS : \n\
       - Un champ de type A : mail.server.net . C'est le nom d'hôte vous permettant ensuite de récupérer vos mails.\n\
       - Un champ de type MX pointant vers le A précédent. C'est en général le nom de domaine porté dans les adresses : machin@domain.net\n\
   Ainsi, lorqu'on vous écrit à machin@domain.net, c'est directement relié à l'hôte mail.server.net" 20 60

    dgetinfo NOMDHOTE "Quel est votre nom de domaine (sans http://)? (ex : mondomaine.com)" "Configuration de postfix"
    dgetinfo DOMAIN "Quel est votre nom d'hote (sans http://)? (ex : smtp.mondomaine.com)"  "Configuration de postfix"
    assign "$1" <<< "dopostfix $NOMDHOTE $DOMAIN"
}

info_sftp() {
    local PORT=""
    local CHROOTDIR=""
    local USER=""
    local HTTPSHARE=""

    # port?
    while ! [[ $PORT =~ ^[0-9]+$ ]]; do
        dialog --title "Quel est le port utilisé pour ssh (22)" --inputbox "port?" 20 60 2> "$tmpfile"
        PORT="$(<"$tmpfile")"
    done
    # répertoire de chroot
    while [ -z "$CHROOTDIR" ]; do
        dialog --title "Emplacement du dossier contenant les documents du sftp (/media/sftp)" --inputbox "dossier?" 20 60 2> "$tmpfile"
        CHROOTDIR="$(<"$tmpfile")"
    done

    # Utilisateurs
    while [ -z "$USER" ]; do
        dialog --title "Quels sont les utilisateurs de ssh (séparés par un espace)" --inputbox "Utilisateurs?" 20 60 2> "$tmpfile"
        USER="$(<"$tmpfile")"
    done

    dialog --title "Configuration de sftp" --yesno "\
    Souhaitez-vous proposer un accès aux documents via un navigateur web? (http://votreserveur.com/~utilisateur)\
        Notez que cela suppose avoir installé nginx avec ce script" 20 60
    HTTPSHARE=$?

    assign "$1" <<< "dosftp $PORT $CHROOTDIR \"$USER\" $HTTPSHARE"
}

info_site() {
    local NOMDHOTE=""
    local ROOTOFHTTP=""
    dgetinfo NOMDHOTE "Quel est votre nom de domaine (sans http://)? (ex : $2.mondomaine.com) " "Configuration de $2"
    dgetinfo ROOTOFHTTP "Dossier contenant votre site? (ex : /srv/www/$2)/ N'oubliez pas le premier /" "Configuration de $2"

    assign "$1" <<< "$NOMDHOTE $ROOTOFHTTP"
}

info_tor() {
    # surnom
    while [ -z "$NICKNAME" ]; do
        dialog --title "Surnom de votre noeud tor pour vous identifier sur le réseau" --inputbox "Surnom?" 20 60 2> "$tmpfile"
        NICKNAME="$(<"$tmpfile")"
    done
    # Adresse de contact
    while [ -z "$EMAIL" ]; do
        dialog --title "Adresse électronique de contact? (bibi@mail.com)" --inputbox "email?" 20 60 2> "$tmpfile"
        EMAIL="$(<"$tmpfile")"
    done
    # pseudo
    while [ -z "$PSEUDO" ]; do
        dialog --title "Votre pseudonyme de contact" --inputbox "Pseudo?" 20 60 2> "$tmpfile"
        PSEUDO="$(<"$tmpfile")"
    done

    while ! [[ $BWRATE =~ ^[0-9]+$ ]]; do
        dialog --title "Bande passante à réserver? (en kB/s)" --inputbox "Bande passante? (en kB/s)" 20 60 2> "$tmpfile"
        BWRATE="$(<"$tmpfile")"
    done
    while ! [[ $BWBURST =~ ^[0-9]+$ ]]; do
        dialog --title "Bande passante maximale à réserver en cas d'éclat? (en kB/s)" --inputbox "Bande passante max? (en kB/s)" 20 60 2> "$tmpfile"
        BWBURST="$(<"$tmpfile")"
    done

    assign "$1" <<< "dotor $NICKNAME $BWRATE $BWBURST $PSEUDO $EMAIL"
}

dialogmenu() {
    # Compatibilité
    export TERM=xterm
    # Commande à éxécuter
    local COMMAND=""
    # fichier temporaire pour menu
    tmpmenu="$(tempfile)"
    trap "rm -f $tmpmenu" EXIT
    # fichier temporaire pour le script
    tmpfile="$(tempfile)"
    trap "rm -f $tmpfile" EXIT
    # fichier temporaire des tâches à réaliser
    tmpwork="$(tempfile)"
    trap "rm -f $tmpwork" EXIT

    dialog --clear\
        --separate-output \
        --backtitle "Host@home - Installez votre serveur à la maison. $VERSION" \
        --checklist 'Tâche(s) à accomplir: \n(<espace> pour cocher, <↑> et <↓> pour déplacer le curseur, <tabulation> pour choisir le bouton, <Entrée> pour valider)' 25 65 18 \
        preparation "Prépare le serveur en le mettant à jour" on\
        securite "sécurité minimale du serveur" on\
        http_nginx "Installer un serveur http(s) (nginx)" off\
        sftp "Partage de fichiers sécurisé avec sftp" off\
        openvpn "Serveur vpn" off\
        postfix "Serveur de courriel postfix" off\
        squirrelmail "Webmail  (squirrelmail)" off\
        rainloop "Webmail (rainloop)" off\
        wallabag "Application de Read-it-later" off\
        owncloud "Hébergez votre nuage" off\
        dropcenter "Déposez et partagez vos documents" off \
        pydio "Partage de fichiers (Ajaxplorer)" off\
        jyraphe "Hébergement et partage de documents" off\
        xmpp "Votre serveur jabber/xmpp" off\
        pluxml "Blog/CMS" off\
        blogotext "Blog" off\
        pico "Un CMS léger utilisant markdown" off\
        kriss "Simple lecteur de flux" off\
        TinyTinyRSS "Lecteur de flux" off\
        mytinytodo "Gestion simple de listes todo" off\
        jotter "Gestion simple de listes todo" off\
        kanboard "Tableau de tâches" off\
        ZeroBin "Pastebin/discussion encrypté" off\
        Shaarli "Clone de delicious" off\
        NoNonsenseForum "Un forum très simple" off\
        fluxBB "Un forum" off\
        phpBB "Un autre forum" off\
        torrent "Seedbox de torrent" off \
        DokuWiki "Simple wiki" off\
        MediaWiki "Wiki complet [complexe]" off\
        tor "Relais tor" off\
        monitorix "Supervisation du serveur" off\
        unbound "[avancé] Serveur DNS" off\
        roundcube "[avancé] Webmail avec sqlite" off\
        roundcube-pgsql "[avancé] Webmail avec postgresql" off 2> "$tmpmenu"
    retval=$?

    case $retval in
        1 | 255) echo "Rien à faire";;
        0) 
            while read line; do
                case "$line" in 
                    "preparation") echo "dopreparation" >> "$tmpwork" ;;
                    "http_nginx") 
                        info_nginx COMMAND 
                        echo "$COMMAND" >> "$tmpwork" 
                        ;;
                    "postfix") 
                        info_postfix COMMAND 
                        echo "$COMMAND" >> "$tmpwork" 
                        ;;
                    "squirrelmail") 
                        info_ssl
                        local NDD=""
                        dgetinfo NDD "Quel est votre nom de domaine (sans http://)? (ex : webmail.mondomaine.com) " "Configuration du webmail"
                        echo "dosquirrelmail $NDD" >> "$tmpwork"
                        ;;
                    "roundcube") 
                        info_ssl
                        info_site COMMAND "roundcube" 
                        echo "doroundcube $COMMAND" >> "$tmpwork"
                        ;;
                    "roundcube-pgsql") 
                        local MDP=""
                        info_ssl
                        info_site COMMAND "roundcube" 
                        dgetinfo MDP "Mot de passe de la base Postgresql?" "Configuration de roundcube"
                        echo "doroundcube-pgsql $COMMAND $MDP" >> "$tmpwork"
                        ;;
                    "rainloop") 
                        info_ssl
                        info_site COMMAND "rainloop" 
                        echo "dorainloop $COMMAND" >> "$tmpwork"
                        ;;
                    "wallabag") 
                        info_ssl
                        info_site COMMAND "wallabag" 
                        echo "dowallabag $COMMAND" >> "$tmpwork"
                        ;;
                    "sftp") 
                        info_sftp COMMAND 
                        echo "$COMMAND" >> "$tmpwork" 
                        ;;
                    "owncloud") 
                        info_ssl
                        info_site COMMAND "owncloud" 
                        echo "doowncloud $COMMAND" >> "$tmpwork"
                        ;;
                    "dotclear") 
                        info_ssl
                        info_site COMMAND "dotclear" 
                        echo "dodotclear $COMMAND" >> "$tmpwork"
                        ;;
                    "xmpp") 
                        dialog --msgbox "Pour que tout fonctionne bien, il serait bon de créer les champs DNS suivants : \n\
                        - Un champ de type A : xmpp.mondomaine.com\n\
                        - Des champs de type SRV vers xmpp.mondomaine.com comme ceci : \n\
                        _xmpp-client._tcp.domaine.net. 18000 IN SRV 0 5 5222 xmpp.domaine.net.\n\
                        _xmpp-server._tcp.domaine.net. 18000 IN SRV 0 5 5269 xmpp.domaine.net."
                        info_ssl
                        local NDD=""
                        dgetinfo NDD "Quel est votre nom de domaine (sans http://)? (ex : xmpp.domaine.com) " "Configuration de xmpp"
                        echo "doxmpp $NDD" >> "$tmpwork"
                        ;;
                    "pluxml")
                        info_ssl
                        info_site COMMAND "pluxml" 
                        echo "dopluxml $COMMAND" >> "$tmpwork"
                        ;;
                    "pico")
                        info_ssl
                        info_site COMMAND "pico" 

                        local PW=''
                        dgetinfo PW "Mot de passe pour pico?" "Choix d'un mot de passe"

                        echo "dopico $COMMAND $PW" >> "$tmpwork"
                        ;;
                    "blogotext") 
                        info_ssl
                        info_site COMMAND "blogotext" 
                        echo "doblogotext $COMMAND" >> "$tmpwork"
                        ;;
                    "kriss") 
                        info_ssl
                        info_site COMMAND "kriss" 
                        echo "dokriss $COMMAND" >> "$tmpwork" ;;
                    "TinyTinyRSS")
                        local MDP=""
                        info_ssl
                        info_site COMMAND "Tiny Tiny RSS" 
                        dgetinfo MDP "Mot de passe de la base Postgresql?" "Configuration de TT-RSS"
                        echo "dottrss $COMMAND $MDP" >> "$tmpwork"
                        ;;
                    "mytinytodo") 
                        info_ssl
                        info_site COMMAND "mytinytodo" 
                        echo "dotinytodo $COMMAND" >> "$tmpwork" ;;
                    "jotter") 
                        info_ssl
                        info_site COMMAND "jotter" 
                        echo "dojotter $COMMAND" >> "$tmpwork" ;;
                    "kanboard") 
                        info_ssl
                        info_site COMMAND "kanboard" 
                        echo "dokanboard $COMMAND" >> "$tmpwork" ;;
                    "tor")
                        info_tor COMMAND
                         echo "$COMMAND" >> "$tmpwork" ;;
                    "dropcenter")
                        info_ssl
                        info_site COMMAND "dropcenter" 
                        echo "dodropcenter $COMMAND" >> "$tmpwork"
                        ;;
                    "ZeroBin")
                        info_site COMMAND "ZeroBin" 
                        echo "dozerobin $COMMAND" >> "$tmpwork"
                        ;;
                    "Shaarli")
                        info_ssl
                        info_site COMMAND "Shaarli" 
                        echo "doshaarli $COMMAND" >> "$tmpwork"
                        ;;
                    "pydio")
                        info_ssl
                        info_site COMMAND "Pydio" 
                        echo "dopydio $COMMAND" >> "$tmpwork"
                        ;;
                    "NoNonsenseForum")
                        info_ssl
                        info_site COMMAND "NoNonsenseForum" 
                        echo "donononsenseforum $COMMAND" >> "$tmpwork"
                        ;;
                    "fluxBB")
                        local MDP=""
                        info_ssl
                        info_site COMMAND "FluxBB" 
                        dgetinfo MDP "Mot de passe de la base Postgresql?" "Configuration de fluxBB"
                        echo "dofluxbb $COMMAND $MDP" >> "$tmpwork"
                        ;;
                    "phpBB")
                        local MDP=""
                        info_ssl
                        info_site COMMAND "phpBB" 
                        dgetinfo MDP "Mot de passe de la base Postgresql?" "Configuration de phpBB"
                        echo "dophpbb $COMMAND $MDP" >> "$tmpwork"
                        ;;
                    "DokuWiki")
                        info_ssl
                        info_site COMMAND "DokuWiki" 
                        echo "dodokuwiki $COMMAND" >> "$tmpwork"
                        ;;
                    "MediaWiki") 
                        info_ssl
                        local NDD=""
                        dgetinfo NDD "Quel est votre nom de domaine (sans http://)? (ex : wiki.mondomaine.com) " "Configuration de mediawiki"
                        echo "domediawiki $NDD" >> "$tmpwork"
                        ;;
                    "torrent")
                        local NDD=""
                        local USERNAME=""
                        local MDP=""
                        info_ssl
                        dgetinfo NDD "Nom de domaine pour rtorrent?" "Configuration de rtorrent"
                        dgetinfo USERNAME "Utilisateur pour rtorrent?" "Configuration de rtorrent"
                        dgetinfo MDP "Mot de passe pour rtorrent?" "Configuration de rtorrent"
                        echo "dotorrent $NDD $USERNAME $MDP" >> "$tmpwork"
                        ;;
                    "jyraphe")
                        local DOCROOT=""
                        info_ssl
                        info_site COMMAND "jyraphe"
                        dgetinfo DOCROOT "Ce dossier doit être en dehors du serveur web"\
                            "Dossier contenant les documents chargés"
                        echo "dojyraphe $COMMAND $DOCROOT" >> "$tmpwork"
                        ;;
                    "unbound")
                        dialog --title "Configuration pour unbound" --yesno "Votre serveur reçoit-il son IP automatiquement avec dhclient? " 20 60
                        local AUTODHCP=$?
                        echo "dounbound $AUTODHCP" >> "$tmpwork"
                        ;;
                    "monitorix")
                        echo "domonitorix" >> "$tmpwork"
                        ;;
                    "openvpn")
                        info_ssl
                        local NDD=""
                        local IP=$(ifconfig | grep 'inet addr:' | grep -v inet6 | grep -vE '127\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}' | cut -d: -f2 | awk '{ print $1}' | head -1)
                        if [[ "$IP" = "" ]]; then
                            IP=$(wget -qO- ipv4.icanhazip.com)
                        fi
                        local PORT=""
                        local CLIENTS=""

                        dgetinfo NDD "Quel est votre nom de domaine (sans http://)? (ex : truc.domaine.org)" "Configuration openvpn"
                        dgetinfo IP "Quelle est l'IP de l'interface sur laquelle vous voulez OpenVPN?" "Configuration openvpn" "$IP"
                        dgetinfo PORT "Quel port utiliser pour OpenVPN?" "Configuration openvpn" "1194"
                        dgetinfo CLIENTS "Quels sont les clients à créer? (ex : client1 client2 client3) (séparés par des espaces)" "Configuration openvpn"
                        echo "doopenvpn $NDD $IP $PORT \"$CLIENTS\"" >> "$tmpwork"
                        ;;

                esac
            done < "$tmpmenu"

            #securité à faire à la fin
            if [ -n "$(grep -o "securite" "$tmpmenu")" ];then
                echo "dosecurite" >> "$tmpwork" 
            fi

            work "$tmpwork" 2>&1 | tee -a "$LOGFILE" 
            dialog --msgbox "Configuration terminée. \n\
Des informations importantes concernant les ports à ouvrir et les mises à jour futures sont présentes dans le fichier $RAPPORT" 20 60
        ;;
    esac
}

if [ "$1" = "-h" ]; then
    showhelp
fi

if [ "$(id -u)" -ne 0 ]; then
    die "Vous devez éxécuter le script avec les droits superutilisateur (root). Essayez 'sudo $0'"
fi

if [ -f "$1" ]; then
    work "$1" 2>&1 | tee -a "$LOGFILE" 
    exit 0
fi

which dialog >/dev/null || (echo "dialog non trouvé -> On l'installe" && installapt dialog)

dialogmenu

exit 0
